File: py/0105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.py
import sys
import heapq
from typing import *
from collections import OrderedDict
from collections import deque


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder or not inorder:
            return None

        root_val = preorder[0]
        root_inorder_idx = inorder.index(root_val)

        root = TreeNode(root_val)
        root.left = self.buildTree(preorder[1: 1 + root_inorder_idx], inorder[0: root_inorder_idx])
        root.right = self.buildTree(preorder[1 + root_inorder_idx:], inorder[root_inorder_idx + 1:])

        return root


if __name__ == "__main__":
    # solution = Solution()

    sys.exit(0)

File: cpp/0105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <queue>
#include <stack>
#include <deque>
#include <list>
#include <map>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* recursion(vector<int>& preorder, vector<int>& inorder, int preStart, int inStart, int inEnd) {
        if (inStart > inEnd) return nullptr;

        TreeNode *root = new TreeNode(preorder[preStart]);
        if (inEnd == inStart) return root;

        int idx = inStart;
        while (inorder[idx] != root->val)
            idx++;

        root->left = recursion(preorder, inorder, preStart + 1, inStart, idx - 1);
        root->right = recursion(preorder, inorder, preStart + 1 + (idx - 1 - inStart) + 1, idx + 1, inEnd);

        return root;
    }


    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return recursion(preorder, inorder, 0, 0, inorder.size() - 1);
    }
};

int main(int argc, char *argv[]) {

    return 0;
}
